### PR TITLE
[WIP] Create new ChainLR.

### PR DESCRIPTION
**Pull request by vincentqb**
The logic for chaining lr scheduler in #21800 may lead to unexpected behavior to the user, see [comment](https://github.com/pytorch/pytorch/issues/22107#issuecomment-506260305). In order to avoid this situation, we decouple the chainable version of a scheduler from its closed form by introducing a new family ChainLR of schedulers, as mentioned [here](https://github.com/pytorch/pytorch/pull/21800#issuecomment-510220723).

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22798 Create new ChainLR.**
* #21800 Revert "Revert "Redefine scheduler to set learning rate using recursive formula" #14010 (#21463)" and enable closed form with non-sequential epoch parameter

Differential Revision: [D16225958](https://our.internmc.facebook.com/intern/diff/D16225958)

### CODE REVIEW

**Comment by ezyang** [comment]
I think it might just be better to squash these two PRs together. I'm happy to review them separately like this though.

File: torch/optim/chain_lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+import math
```
Comment (ezyang) [inline torch/optim/chain_lr_scheduler.py]: Since there is a new module, you will want to add a module-level docblock, and an entry for the module in the docs, in `docs/source/optim.rst`

File: torch/optim/chain_lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+from .lr_scheduler import LambdaLR, StepLR, MultiStepLR, ExponentialLR, CosineAnnealingLR
+
+
+class LambdaChainLR(LambdaLR):
+    """Sets the learning rate of each parameter group to the initial lr
```
Comment (ezyang) [inline torch/optim/chain_lr_scheduler.py]: Now that we've split the ChainLRs into their own modules, it's worth explicitly rewriting the docs to be more clear about what exactly they do (and how they differ from the regular LR classes). Specifically, (1) there should probably be some module level overall description of how chain LRs work, and (2) we can probably drop the closed form equations from these classes (and instead just reference their corresponding non-closed version as a reference.)

File: torch/optim/chain_lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
+import math
+
+from .lr_scheduler import LambdaLR, StepLR, MultiStepLR, ExponentialLR, CosineAnnealingLR
+
+
+class LambdaChainLR(LambdaLR):
```
Comment (ezyang) [inline torch/optim/chain_lr_scheduler.py]: This isn't really an "is-a" situation, but I guess Python doesn't really give us any way to actually do private inheritance lol.

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -252,17 +230,12 @@ class MultiStepLR(_LRScheduler):
     """
 
     def __init__(self, optimizer, milestones, gamma=0.1, last_epoch=-1):
-        self.milestones = Counter(milestones)
+        self.milestones = milestones
+        self._milestones_counter = Counter(milestones)
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: It's probably worth a comment here that this Counter is here specifically for a subclass instance (which lives in another file).

Actually, this is one of the reasons why I preferred the version of the code where the closed-form and recursive formulas all lived in one class, it means that all of the formulas for a given scheduler were in one place and you wouldn't easily forget to update one if you updated the other (and similarly, if you added a new LR scheduler, you would remember to add another LR scheduler.) Maybe "chainability" should still be toggled as a keyword argument, rather than blasting out an entirely separate module hierarchy?

File: torch/optim/__init__.pyi [hunk 5]
```diff
@@ -1,3 +0,0 @@
-from .sgd import SGD as SGD
-from .adam import Adam as Adam
-from . import lr_scheduler as lr_scheduler
```
Comment (ezyang) [inline torch/optim/__init__.pyi]: I don't think you meant to delete this?

File: test/test_optim.py [hunk 6]
```diff
@@ -1288,7 +1287,6 @@ def _test_get_lr(self, schedulers, targets, epochs=10):
             [scheduler.step(epoch) for scheduler in schedulers]
             result = [scheduler.get_lr() for scheduler in schedulers]
             target = [[t[epoch] for t in targets]] * len(schedulers)
-            # print(target)
```
Comment (ezyang) [inline test/test_optim.py]: kill this

**Review by ezyang — CHANGES_REQUESTED** [review]
I think the main problem in this PR is that the docs are insufficient. I also made some suggestions about alternative API formulations (and my reasoning behind them), which are worth considering (although by no means a hard request--accept this suggestion at your discretion!) You should probably figure out if you want to take my API suggestion or not first, since that will affect how you write docs.

**Comment by vincentqb** [comment]
> I think it might just be better to squash these two PRs together. I'm happy to review them separately like this though.

I kept the two separate mostly because I wanted to keep #21800 intact and offer this one as a preliminary alternative of what ChainLR would look like.

**Comment by vincentqb** [comment]
Following discussion in #21800, we are going in a different direction.
